### PR TITLE
Lido app/marcus/maxa

### PR DESCRIPTION
**Pull request by marhong2**
remade to hide the irrelevant changes

### CODE REVIEW

File: formulas/functions/statistical.js [hunk 1]
```diff
@@ -215,7 +215,23 @@ const StatisticalFunctions = {
...
+      if(a === true)
+        return 1
+      if(!isNaN(parseFloat(a)))
+        return parseFloat(a)
+      if(!isNaN(parseInt(a)))
+        return parseInt(a)
+      return 0
+    })
+    if(updatedArr.length === 0)
+      return 0;
```
Comment (jmorganmartin) [inline formulas/functions/statistical.js:232]: This check could be done on `flatArr`,  before the map, since `flatArr` and `updatedArr` will always be the same length with that map function.  Also, I'm not sure if this should return `0` or throw an error (`N/A`?) in that case.

File: formulas/functions/statistical.js [hunk 2]
```diff
@@ -215,7 +215,23 @@ const StatisticalFunctions = {
 
   MAX: () => {},
 
-  MAXA: () => {},
+  MAXA: (...values) => {
+    const arr = [];
+    arr.push(...values.map(a => a.value))
```
Comment (jmorganmartin) [inline formulas/functions/statistical.js:220]: `Array.map` returns a new array already. Spreading it (`...values`) just to `push()` the result is unneccessary. 

```suggestion
    const arr = values.map(a => a.value);
```

File: formulas/functions/statistical.js [hunk 3]
```diff
@@ -215,7 +215,23 @@ const StatisticalFunctions = {
...
+        return 1
+      if(!isNaN(parseFloat(a)))
+        return parseFloat(a)
+      if(!isNaN(parseInt(a)))
+        return parseInt(a)
+      return 0
+    })
+    if(updatedArr.length === 0)
+      return 0;
+    return Math.max(...updatedArr)
```
Comment (jmorganmartin) [inline formulas/functions/statistical.js:233]: Getting very "edge-case" here, but spreading an array into `Math.max()` (and similar functions, like `push()` above) can fail if there are too many elements in the array (you can end up hitting the JS parameter limit). 

It's safer to use a loop (or something that loops, like `Array.reduce()`)

```suggestion
    return updatedArr.reduce((a, b) => {
      return Math.max(a, b);
    }, -Infinity);
```

See: 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/max

(note, the example they give with `reduce()` ^^ would actually fail if the array contained all negative numbers, since it starts with `0` instead of `-Infinity`.)

Which links to:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply#using_apply_and_built-in_functions

File: formulas/functions/statistical.js [hunk 4]
```diff
@@ -215,7 +215,23 @@ const StatisticalFunctions = {
...
+    arr.push(...values.map(a => a.value))
+    const flatArr = FormulaHelpers.flattenDeep(arr)
+    const updatedArr = flatArr.map(a => {
+      if(a === true)
+        return 1
+      if(!isNaN(parseFloat(a)))
+        return parseFloat(a)
+      if(!isNaN(parseInt(a)))
+        return parseInt(a)
+      return 0
```
Comment (jmorganmartin) [inline formulas/functions/statistical.js:229]: If the data we are checking contains _all_ negative numbers, this line will cause the function to return `0`, which would not be the max.

```suggestion
      return -Infinity
```

**Review by jmorganmartin — CHANGES_REQUESTED** [review]
Requesting changes for a few edge-cases that could cause bugs.

Probably worth adding some tests to confirm proper behavior in those edge cases

File: formulas/functions/statistical.js (hunk 3 above)
Comment (jmorganmartin) [reply formulas/functions/statistical.js:233]: Update (sorry these comments are a bit contradictory at this point. I went and ate lunch and couldn't stop thinking about this, haha):

If we use `reduce` above to coerce the `inputArr` to appropriate numbers (and remove `NaN`s), then we should probably do the max comparison in the same loop, rather than running two loops.

Something like:

```js
let numCount = 0;
const max = flatArr.reduce((a, b) => {
  // coerce to number, increment numCount if !isNaN
  
  // compare a to b, return larger
}, -Infinity);
// return max if numCount > 0, else 0 or Error
```

File: formulas/functions/statistical.js (hunk 4 above)
Comment (marhong2) [reply formulas/functions/statistical.js:229]: I changed it a little, Instead of returning a number, I return null if its an invalid input then filter out all the nulls

File: formulas/functions/statistical.js (hunk 1 above)
Comment (marhong2) [reply formulas/functions/statistical.js:232]: Excel returns 0

File: formulas/functions/statistical.js (hunk 3 above)
Comment (marhong2) [reply formulas/functions/statistical.js:233]: Since I filter out all of the nulls earlier, I think I can just run a single reduce loop

File: formulas/functions/statistical.js (hunk 4 above)
Comment (marhong2) [reply formulas/functions/statistical.js]: const updatedArr = flatArr.map(a => {
      if(a === true)
        return 1
      if(!isNaN(parseFloat(a)))
        return parseFloat(a)
      if(!isNaN(parseInt(a)))
        return parseInt(a)
      return null
    })
    const finalArr = updatedArr.filter((a) => a != null)

File: formulas/functions/statistical.js (hunk 3 above)
Comment (marhong2) [reply formulas/functions/statistical.js]: return finalArr.reduce((a,b) => Math.max(a,b))

Comment (marhong2) [reply formulas/functions/statistical.js]: Also I take care of the case when finalArr is of len 0 earlier, so I dont think a default is nessessary

**Review by tyleradams — APPROVED** [review]
